gpjax/dataset.py

```diff
@@ -15,30 +15,55 @@
 
 from dataclasses import dataclass
 
-from beartype.typing import Optional
+from beartype.typing import (
+    Optional,
+    Union,
+)
 import jax.numpy as jnp
-from jaxtyping import Num
+from jaxtyping import (
+    Bool,
+    Num,
+)
 from simple_pytree import Pytree
 
 from gpjax.typing import Array
 
 
+class _Missing:
+    """Sentinel class for not-yet-computed mask"""
+
+
 @dataclass
 class Dataset(Pytree):
     r"""Base class for datasets.
 
     Attributes
     ----------
-        X (Optional[Num[Array, "N D"]]): Input data.
-        y (Optional[Num[Array, "N Q"]]): Output data.
+        X: Num[Array, "N D"], optional:
+            Input data.
+        y: Num[Array, "N Q"], optional:
+            Output data.
+        mask: Bool[Array, "N Q"] or None, optional
+            Mask for the output data. By default, a mask will be computed based on the output data.
+            User can optionally specify a pre-computed mask, or pass `None` which means no mask
+            will be computed.
     """
 
     X: Optional[Num[Array, "N D"]] = None
     y: Optional[Num[Array, "N Q"]] = None
+    mask: Optional[Union[Bool[Array, "N Q"], None]] = _Missing()
 
     def __post_init__(self) -> None:
         r"""Checks that the shapes of $`X`$ and $`y`$ are compatible."""
         _check_shape(self.X, self.y)
+        if isinstance(self.mask, _Missing):
+            if self.y is not None:
+                if jnp.any(mask := jnp.isnan(self.y)):
```

Comment:
I find
```
mask = jnp.isnan(self.y)
if jnp.any(mask)
```
easier to read than the := operator. this hold for later as well but is mostly a question of style.

Reply:
The walrus operator is still a controversial feature after all 😅 I agree it's not really needed in this case but later it makes for a cleaner code imo

Reply:
I think I agree with Ingmar here.